tools/walletextension/readwriter/read_writer.go

```diff
@@ -0,0 +1,102 @@
+package readwriter
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/obscuronet/go-obscuro/go/common/log"
+
+	"github.com/gorilla/websocket"
+)
+
+const (
+	headerUpgrade   = "Upgrade"
+	headerWebsocket = "websocket"
+	httpCodeErr     = 500
+)
+
+var upgrader = websocket.Upgrader{} // Used to upgrade connections to websocket connections.
+
+// ReadWriter handles reading and writing Ethereum JSON RPC requests.
+type ReadWriter interface {
+	ReadRequest() ([]byte, error)
+	WriteResponse(msg []byte) error
+	HandleError(msg string)
+	SupportsSubscriptions() bool
+}
+
+// HTTPReadWriter is a ReadWriter over HTTP.
+type HTTPReadWriter struct {
+	resp http.ResponseWriter
+	req  *http.Request
+}
+
+// WSReadWriter is a ReadWriter over websockets.
+type WSReadWriter struct {
+	conn *websocket.Conn
+}
+
+func NewReadWriter(resp http.ResponseWriter, req *http.Request) (ReadWriter, error) {
+	for _, header := range req.Header[headerUpgrade] {
+		if header == headerWebsocket {
+			conn, err := upgrader.Upgrade(resp, req, nil)
+			if err != nil {
+				err = fmt.Errorf("unable to upgrade to websocket connection")
+				logAndSendErr(resp, err.Error())
+				return nil, err
+			}
+			return &WSReadWriter{
+				conn: conn,
+			}, nil
+		}
+	}
+
+	return &HTTPReadWriter{resp: resp, req: req}, nil
+}
+
+func (h HTTPReadWriter) ReadRequest() ([]byte, error) {
```

Comment:
Should these be instance methods ?
`func (h HTTPReadWriter) ReadRequest() ([]byte, error)` to `func (h *HTTPReadWriter) `

Reply:
Fixed.